File: Data Structures and Algorithms/20-ValidParentheses/20-ValidParentheses.py
# Last updated: 6/4/2026, 6:37:48 PM
class Solution:
    def isValid(self, s: str) -> bool:
        chars = {'}':'{',']':'[',')':'('}
        stack = []
        for c in s:
            if c in chars:
                if stack and stack[-1] == chars[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return True if not stack else False
        

File: Data Structures and Algorithms/20-ValidParentheses/20-ValidParentheses.cpp
// Last updated: 6/4/2026, 6:37:39 PM
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char,char> hashMap = {
            {'}','{'},
            {']','['},
            {')','('}
            };
        stack<int> Stack;
        for(int i=0; i < s.length(); i++)
        {
            if(hashMap.find(s[i]) != hashMap.end())
            {
                if (!Stack.empty() && Stack.top() == hashMap[s[i]])
                    Stack.pop();
                else
                    return false;
            }
            else{
                Stack.push(s[i]);
            }
        }
        return true ? Stack.empty() : false;
    }
};